cartconditions/Coupon.php

```diff
@@ -34,51 +38,78 @@ public function getValue()
 
     public function getModel()
     {
-        return $this->couponModel;
+        if (!strlen($couponCode = $this->getMetaData('code')))
+            return self::$couponModel;
+
+        if (is_null(self::$couponModel))
+            self::$couponModel = Coupons_model::getByCode($couponCode);
+
+        if (self::$couponModel AND strtolower(self::$couponModel->code) !== strtolower($couponCode))
+            self::$couponModel = Coupons_model::getByCode($couponCode);
+
+        return self::$couponModel;
     }
 
-    public function beforeApply()
+    public function getApplicableItems($couponModel)
     {
-        if (!strlen($couponCode = $this->getMetaData('code')))
-            return FALSE;
+        $applicableItems = $couponModel->menus->pluck('menu_id');
+        $couponModel->categories->pluck('category_id')
+            ->each(function ($category) use (&$applicableItems) {
+                $applicableItems = $applicableItems
+                    ->merge(Menus_model::whereHasCategory($category)->pluck('menu_id'));
+            });
 
-        if (is_null($this->couponModel))
-            $this->couponModel = Coupons_model::getByCode($couponCode);
+        self::$applicableItems = $applicableItems;
+
+        return self::$applicableItems;
+    }
+
+    public function onLoad()
+    {
+        if (!strlen($couponCode = $this->getMetaData('code')))
+            return;
 
         try {
-            if (!$this->couponModel)
+            if (!$couponModel = $this->getModel())
                 throw new ApplicationException(lang('igniter.cart::default.alert_coupon_invalid'));
 
-            $this->validateCoupon();
+            $this->validateCoupon($couponModel);
+
+            $this->getApplicableItems($couponModel);
         }
         catch (Exception $ex) {
             flash()->alert($ex->getMessage())->now();
             $this->removeMetaData('code');
+        }
+    }
 
+    public function beforeApply()
+    {
+        $couponModel = $this->getModel();
+        if (!$couponModel OR !$couponModel->is_limited_to_cart_item)
```

Comment:
This should probably change to `if (!$couponModel OR $couponModel->is_limited_to_cart_item)`